### DISCUSSION TITLE
modes/benchmark/

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#9137] [Docs]
# modes/benchmark/

Learn how to profile speed and accuracy of YOLOv8 across various export formats; get insights on mAP50-95, accuracy_top5 metrics, and more.

https://docs.ultralytics.com/modes/benchmark/

**Comment by mansoorshah007** [comment]
Hi, I want to run inferene on video.mp4 witht he below code:
engine    = YOLO('yolov8x.engine')
inference = engine.track(source="/content/output-video.mp4",  save = True) 
and I am encountering error: 

AttributeError                            Traceback (most recent call last)

<ipython-input-7-c0663e08864a> in <cell line: 2>()
      1 engine    = YOLO('/content/drive/MyDrive/Projects/Horse Tracking/yolov8x.engine')
----> 2 inference = engine.track(source="/content/output-video.mp4", conf = 0.3, iou = 0.5, show = True, save = True)

4 frames

/usr/local/lib/python3.10/dist-packages/ultralytics/nn/autobackend.py in __init__(self, weights, device, dnn, data, fp16, batch, fuse, verbose)
    234                 metadata = json.loads(f.read(meta_len).decode("utf-8"))  # read metadata
    235                 model = runtime.deserialize_cuda_engine(f.read())  # read engine
--> 236             context = model.create_execution_context()
    237             bindings = OrderedDict()
    238             output_names = []

AttributeError: 'NoneType' object has no attribute 'create_execution_context'

**Reply by pderrenger** [reply]
@mansoorshah007 hello! It looks like you're encountering an issue with the `.create_execution_context()` method while trying to run the tracker on a video. This error typically happens when YOLO is unable to properly load the engine file, resulting in a `None` model object.

One common reason for this could be an incorrect path to your engine file or issues with the engine file itself. Let's take a closer look at how you're loading the model:

```python
engine = YOLO('yolov8x.engine')
```

Please ensure that the path to `'yolov8x.engine'` is correct and that the file is accessible. If you're running this in a Google Colab notebook, double-check that the drive is mounted correctly and your path starts with `/content/drive/My Drive/...` if the file is in your Google Drive.

If the path is correct and you're still facing issues, it might be worth trying to re-export your model to an engine file to ensure it's not corrupted. Here’s a quick reminder on how to do it:

```bash
yolo export model=path/to/yolov8n.pt format=engine
```

If the problem persists, please feel free to share more details about your environment (e.g., PyTorch version, CUDA version) so we can help you troubleshoot further. 😊

**Comment by Zency-Sun** [comment]
Hi, what should I do if I want to benchmark using only Pytorch, ONNX, openvino and tensorRT?

**Reply by pderrenger** [reply]
Hey there! 😊 To benchmark using only PyTorch, ONNX, OpenVINO, and TensorRT, you can specify the formats you're interested in when running the benchmark command. Here's a concise way to do it:

For Python, you can run separate `benchmark` commands for each format after exporting your model to that format. For example:

```python
from ultralytics import YOLO, benchmark

model = YOLO('yolov8n.pt')  # Load your model
data = 'coco128.yaml'  # Define your dataset

# Ensure you've exported your model to the desired formats beforehand
formats = ['torchscript', 'onnx', 'openvino', 'engine']  # PyTorch, ONNX, OpenVINO, TensorRT formats
for fmt in formats:
    model.export(format=fmt)  # Export to each format
    benchmark(model=f'yolov8n.{fmt if fmt != "engine" else "engine"}', data=data, imgsz=640, device=0)
```

And for CLI, you'd first export your model to each format and then run benchmarks separately:

```bash
# Assuming your model is already exported to each format
yolo benchmark model=yolov8n.pt data='coco128.yaml' imgsz=640 device=0
yolo benchmark model=yolov8n.onnx data='coco128.yaml' imgsz=640 device=0
yolo benchmark model=yolov8n_openvino_model/ data='coco128.yaml' imgsz=640 device=0
yolo benchmark model=yolov8n.engine data='coco128.yaml' imgsz=640 device=0
```

Make sure your model is exported to each format before benchmarking. This approach allows you to directly compare the performance across the specified formats. Hope this helps! If there are more questions, feel free to ask.

**Reply by Zency-Sun** [reply]
when I use the under code to deploy：
'    model = YOLO("yolov8s.engine")
    metrics = model.val(data="ultralytics/cfg/datasets/coco128.yaml",device='cuda:0')   '
but I meet the error,What should I do?，Thanks your reply！！！

Traceback (most recent call last):
  File "/media/zency/SSD-Data/Codes/yolov8_v8.1.0/ultralytics/nn/autobackend.py", line 220, in __init__
    import tensorrt as trt  # noqa https://developer.nvidia.com/nvidia-tensorrt-download
  File "/home/zency/anaconda3/envs/yolov8plus/lib/python3.9/site-packages/tensorrt/__init__.py", line 68, in <module>
    from .tensorrt import *
ImportError: libnvinfer.so.8: cannot open shared object file: No such file or directory

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/media/zency/SSD-Data/Codes/yolov8_v8.1.0/run_test.py", line 190, in <module>
    yolo_mytest()
  File "/media/zency/SSD-Data/Codes/yolov8_v8.1.0/run_test.py", line 116, in yolo_mytest
    metrics = model.val(data="ultralytics/cfg/datasets/coco128.yaml",device='cuda:0')   # cpu   cuda:0
  File "/media/zency/SSD-Data/Codes/yolov8_v8.1.0/ultralytics/engine/model.py", line 513, in val
    validator(model=self.model)
  File "/home/zency/anaconda3/envs/yolov8plus/lib/python3.9/site-packages/torch/autograd/grad_mode.py", line 27, in decorate_context
    return func(*args, **kwargs)
  File "/media/zency/SSD-Data/Codes/yolov8_v8.1.0/ultralytics/engine/validator.py", line 124, in __call__
    model = AutoBackend(
  File "/home/zency/anaconda3/envs/yolov8plus/lib/python3.9/site-packages/torch/autograd/grad_mode.py", line 27, in decorate_context
    return func(*args, **kwargs)
  File "/media/zency/SSD-Data/Codes/yolov8_v8.1.0/ultralytics/nn/autobackend.py", line 224, in __init__
    import tensorrt as trt  # noqa
  File "/home/zency/anaconda3/envs/yolov8plus/lib/python3.9/site-packages/tensorrt/__init__.py", line 68, in <module>
    from .tensorrt import *
ImportError: libnvinfer.so.8: cannot open shared object file: No such file or directory
Ultralytics YOLOv8.1.24 🚀 Python-3.9.18 torch-1.12.1 CUDA:0 (NVIDIA GeForce RTX 3090, 24268MiB)
Loading yolov8s.engine for TensorRT inference...

**Reply by Zency-Sun** [reply]
I solved the problem above, but there is a new problem:

[TRT] [W] CUDA initialization failure with error: 35

**Reply by pderrenger** [reply]
Hey there! 🌟 It looks like you've encountered the CUDA initialization failure error with code 35, which usually indicates a CUDA driver/runtime version mismatch or an issue with accessing your GPU.

Here are a couple of steps you can try to resolve the issue:
1. **Check CUDA Compatibility**: Ensure that the installed CUDA version is compatible with the version required by TensorRT.
2. **Update Drivers**: Sometimes, simply updating your NVIDIA drivers can resolve such issues.
3. **Environment Check**: Make sure your environment variables (like `PATH` and `LD_LIBRARY_PATH` for Linux) are correctly set to include paths to CUDA toolkit and libraries.

If these steps don't help, could you provide a bit more context about your setup? Specifically:
- The versions of CUDA, TensorRT, and PyTorch you're using.
- The output of `nvidia-smi` to check if the GPU is being recognized properly.

This information can help pinpoint the issue more accurately. Thanks for reaching out, and looking forward to getting you past this hurdle! 🚀

**Reply by Zency-Sun** [reply]
Thank you for your patient reply, I have resolved the above issue!

**Reply by pderrenger** [reply]
@Zency-Sun i'm glad to hear you've resolved your previous issue! Regarding your question about model benchmarking, it's a great way to understand your model's performance across different formats. Here's a quick overview of what you might find useful:

- **Why Benchmark?**: Helps you make informed decisions by understanding the trade-off between speed and accuracy across different formats and on varied hardware.
- **Key Metrics**: Look at mAP50-95 for detection accuracy, and inference time for speed. This gives you a holistic view of how well your model performs.

For benchmarking specific formats like PyTorch, ONNX, OpenVINO, and TensorRT, you can use the `benchmark` utility. Here's a Python snippet to get you started:

```python
from ultralytics import YOLO, benchmark

model = YOLO('yolov8n.pt')  # Load your model
data = 'coco128.yaml'  # Define your dataset

# Benchmark across specific formats
for format in ['torchscript', 'onnx', 'openvino', 'engine']:
    model.export(format=format)  # Export to the format
    benchmark(model=f'yolov8n.{format}', data=data, imgsz=640, device='cuda:0')  # Run benchmark
```

And for CLI, make sure you've exported to each format first:

```bash
yolo export model=yolov8n.pt format=onnx
yolo benchmark model=yolov8n.onnx data='coco8.yaml' imgsz=640 device=0
# Repeat for 'torchscript', 'openvino', and 'engine'
```

Remember, exporting to ONNX or OpenVINO can speed up CPU inference, and TensorRT is excellent for maximizing GPU efficiency. 🚀

Let me know if there's anything else you need help with!

**Comment by bayda369** [comment]
Hi, I want to customize the yolov8 to detect and track only the objects determined by the user, like selecting 3 persons in different positions out of 10 persons in the video appears in the first frame of the video and ignore any other persons, what are your suggestions?

**Reply by pderrenger** [reply]
@bayda369 hello,

For customizing YOLOv8 to detect and track specific objects (like selecting certain individuals from a video), you can approach this by integrating a pre-processing step to identify and tag the objects of interest in the first frame. Here’s a brief outline on how you might proceed:

1. **Object Detection**: Use YOLOv8 to detect all persons in the first frame.
2. **Object Selection**: Implement a method to select the desired objects. This could be manual selection via bounding box coordinates or an automated method based on certain criteria.
3. **Tracking**: Once the objects are tagged, you can use the tracking functionality to follow these objects across frames. You might need to modify the tracking logic to focus only on the tagged objects.

For the tracking part, you can utilize the existing tracking capabilities of YOLOv8, which supports multiple tracking algorithms. You may need to customize the tracking code to ensure it only tracks the initially selected objects.

This approach requires some custom coding, especially for the selection and modified tracking logic. If you need further guidance on the specific code modifications, feel free to ask!

**Reply by bayda369** [reply]
yes please, I need further guidance on the specific code modifications

**Reply by pderrenger** [reply]
Hello @bayda369,

Absolutely, I'd be happy to guide you through the specific code modifications needed to achieve your goal of detecting and tracking only selected objects in a video.

### Step-by-Step Guide

1. **Initial Object Detection**:
   First, detect all persons in the initial frame using YOLOv8.

2. **Object Selection**:
   Implement a method to manually or programmatically select the desired persons from the detected objects.

3. **Tracking**:
   Modify the tracking logic to focus only on the selected persons.

### Code Example

Here's a Python script that demonstrates these steps:

```python
import cv2
from ultralytics import YOLO

# Load the YOLOv8 model
model = YOLO("yolov8n.pt")

# Open the video file
video_path = "path/to/your/video/file.mp4"
cap = cv2.VideoCapture(video_path)

# Read the first frame
ret, frame = cap.read()
if not ret:
    print("Failed to read the video")
    cap.release()
    exit()

# Detect all persons in the first frame
results = model(frame)
boxes = results[0].boxes.xyxy.cpu().numpy()  # Get bounding boxes
selected_boxes = []  # List to store selected boxes

# Display the first frame with bounding boxes
for box in boxes:
    x1, y1, x2, y2 = map(int, box[:4])
    cv2.rectangle(frame, (x1, y1), (x2, y2), (0, 255, 0), 2)

cv2.imshow("Select Persons", frame)
cv2.waitKey(0)

# Manually select persons (for simplicity, we assume user clicks on the bounding boxes)
# Here, you can implement a more sophisticated selection method if needed
selected_boxes = boxes[:3]  # Example: selecting the first 3 persons

# Tracking the selected persons
trackers = cv2.MultiTracker_create()
for box in selected_boxes:
    x1, y1, x2, y2 = map(int, box[:4])
    tracker = cv2.TrackerKCF_create()  # You can choose different tracker types
    trackers.add(tracker, frame, (x1, y1, x2 - x1, y2 - y1))

# Loop through the video frames
while cap.isOpened():
    ret, frame = cap.read()
    if not ret:
        break

    # Update the trackers
    success, boxes = trackers.update(frame)
    if success:
        for i, new_box in enumerate(boxes):
            x, y, w, h = map(int, new_box)
            cv2.rectangle(frame, (x, y), (x + w, y + h), (0, 255, 0), 2)
    else:
        print("Tracking failed")

    # Display the frame
    cv2.imshow("Tracking", frame)
    if cv2.waitKey(1) & 0xFF == ord('q'):
        break

cap.release()
cv2.destroyAllWindows()
```

### Explanation

1. **Initial Detection**:
   - The first frame is read and all persons are detected using YOLOv8.
   - Bounding boxes are drawn around detected persons.

2. **Object Selection**:
   - For simplicity, the first 3 detected persons are selected. You can replace this with a more interactive selection method.

3. **Tracking**:
   - OpenCV's `MultiTracker` is used to track the selected persons across subsequent frames.

Feel free to customize the selection method and tracking algorithm as per your requirements. If you need further assistance, don't hesitate to ask! 😊

**Comment by WooFy11** [comment]
# @title Speed Tracker
import cv2
import numpy as np
from ultralytics import YOLO

from ultralytics.utils.checks import check_imshow
from ultralytics.utils.plotting import Annotator, colors

from collections import defaultdict

track_history = defaultdict(lambda: [])
model = YOLO("/content/drive/MyDrive/DATASET_TUGASAKHIR/PlayerResults/weights/best.pt")
names = model.model.names
video_path = "/content/drive/MyDrive/DATASET_TUGASAKHIR/testvid/BDEvid1.mp4"
cap = cv2.VideoCapture(video_path)
assert cap.isOpened(), "Error reading video file"

w, h, fps = (int(cap.get(x)) for x in (cv2.CAP_PROP_FRAME_WIDTH, cv2.CAP_PROP_FRAME_HEIGHT, cv2.CAP_PROP_FPS))

result = cv2.VideoWriter("SpeedTracker.avi",
                       cv2.VideoWriter_fourcc(*'mp4v'),
                       fps,
                       (w, h))

time_per_frame = 1.0 / fps
real_world_scale = 0.05  # Example: 0.05 meters per pixel

while cap.isOpened():
    success, frame = cap.read()
    if success:
        results = model.track(frame, persist=True, verbose=False, tracker="/content/drive/MyDrive/DATASET_TUGASAKHIR/Tracker/bytetrack.yaml")
        boxes = results[0].boxes.xyxy.cpu()

        if results[0].boxes.id is not None:

            # Extract prediction results
            clss = results[0].boxes.cls.cpu().tolist()
            track_ids = results[0].boxes.id.int().cpu().tolist()
            confs = results[0].boxes.conf.float().cpu().tolist()

            # Annotator Init
            annotator = Annotator(frame, line_width=2)

            for box, cls, track_id in zip(boxes, clss, track_ids):
                # Store tracking history
                track = track_history[track_id]
                track.append((int((box[0] + box[2]) / 2), int((box[1] + box[3]) / 2)))
                if len(track) > 30:
                    track.pop(0)

                if len(track) >= 2:
                    # Calculate pixel distance between the last and second to last points
                    dx = track[-1][0] - track[-2][0]
                    dy = track[-1][1] - track[-2][1]
                    pixel_distance = (dx ** 2 + dy ** 2) ** 0.5
                    real_distance = pixel_distance * real_world_scale
                    speed = real_distance / time_per_frame  # speed in meters per second
                    label = f"{names[int(cls)]} {track_id} {speed:.2f} m/s"
                    annotator.box_label(box, color=colors(int(cls), True), label=label)

                    # Plot tracks
                    points = np.array(track, dtype=np.int32).reshape((-1, 1, 2))
                    cv2.circle(frame, (track[-1]), 7, colors(int(cls), True), -1)
                    cv2.polylines(frame, [points], isClosed=False, color=colors(int(cls), True), thickness=2)
                else:
                    label = f"{names[int(cls)]} {track_id}"
                    annotator.box_label(box, color=colors(int(cls), True), label=label)

        result.write(frame)
        if cv2.waitKey(1) & 0xFF == ord("q"):
            break
    else:
        break

result.release()
cap.release()
cv2.destroyAllWindows()


how do we benchmark the bytetrack perfomance per player ID?

**Comment by WooFy11** [comment]
# @title Speed Tracker
import cv2
import numpy as np
from ultralytics import YOLO

from ultralytics.utils.checks import check_imshow
from ultralytics.utils.plotting import Annotator, colors

from collections import defaultdict

track_history = defaultdict(lambda: [])
model = YOLO("/content/drive/MyDrive/DATASET_TUGASAKHIR/PlayerResults/weights/best.pt")
names = model.model.names
video_path = "/content/drive/MyDrive/DATASET_TUGASAKHIR/testvid/BDEvid1.mp4"
cap = cv2.VideoCapture(video_path)
assert cap.isOpened(), "Error reading video file"

w, h, fps = (int(cap.get(x)) for x in (cv2.CAP_PROP_FRAME_WIDTH, cv2.CAP_PROP_FRAME_HEIGHT, cv2.CAP_PROP_FPS))

result = cv2.VideoWriter("SpeedTracker.avi",
                       cv2.VideoWriter_fourcc(*'mp4v'),
                       fps,
                       (w, h))

time_per_frame = 1.0 / fps
real_world_scale = 0.05  # Example: 0.05 meters per pixel

while cap.isOpened():
    success, frame = cap.read()
    if success:
        results = model.track(frame, persist=True, verbose=False, tracker="/content/drive/MyDrive/DATASET_TUGASAKHIR/Tracker/bytetrack.yaml")
        boxes = results[0].boxes.xyxy.cpu()

        if results[0].boxes.id is not None:

            # Extract prediction results
            clss = results[0].boxes.cls.cpu().tolist()
            track_ids = results[0].boxes.id.int().cpu().tolist()
            confs = results[0].boxes.conf.float().cpu().tolist()

            # Annotator Init
            annotator = Annotator(frame, line_width=2)

            for box, cls, track_id in zip(boxes, clss, track_ids):
                # Store tracking history
                track = track_history[track_id]
                track.append((int((box[0] + box[2]) / 2), int((box[1] + box[3]) / 2)))
                if len(track) > 30:
                    track.pop(0)

                if len(track) >= 2:
                    # Calculate pixel distance between the last and second to last points
                    dx = track[-1][0] - track[-2][0]
                    dy = track[-1][1] - track[-2][1]
                    pixel_distance = (dx ** 2 + dy ** 2) ** 0.5
                    real_distance = pixel_distance * real_world_scale
                    speed = real_distance / time_per_frame  # speed in meters per second
                    label = f"{names[int(cls)]} {track_id} {speed:.2f} m/s"
                    annotator.box_label(box, color=colors(int(cls), True), label=label)

                    # Plot tracks
                    points = np.array(track, dtype=np.int32).reshape((-1, 1, 2))
                    cv2.circle(frame, (track[-1]), 7, colors(int(cls), True), -1)
                    cv2.polylines(frame, [points], isClosed=False, color=colors(int(cls), True), thickness=2)
                else:
                    label = f"{names[int(cls)]} {track_id}"
                    annotator.box_label(box, color=colors(int(cls), True), label=label)

        result.write(frame)
        if cv2.waitKey(1) & 0xFF == ord("q"):
            break
    else:
        break

result.release()
cap.release()
cv2.destroyAllWindows()

how to collect the ground truth ID's from this code

**Comment by themeowsketeer** [comment]
Greetings,
I'm facing the following error when trying to benchmark using trained weight from RT-DeTR-L model, which is known to be fully supported by Ultralytics:
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
File /opt/conda/lib/python3.10/site-packages/ultralytics/utils/benchmarks.py:133, in benchmark(model, data, imgsz, half, int8, device, verbose)
    132 assert i != 5 or platform.system() == "Darwin", "inference only supported on macOS>=10.13"  # CoreML
--> 133 exported_model.predict(ASSETS / "bus.jpg", imgsz=imgsz, device=device, half=half)
    135 # Validate

File /opt/conda/lib/python3.10/site-packages/ultralytics/engine/model.py:563, in Model.predict(self, source, stream, predictor, **kwargs)
    562     self.predictor.set_prompts(prompts)
--> 563 return self.predictor.predict_cli(source=source) if is_cli else self.predictor(source=source, stream=stream)

File /opt/conda/lib/python3.10/site-packages/ultralytics/engine/predictor.py:168, in BasePredictor.__call__(self, source, model, stream, *args, **kwargs)
    167 else:
--> 168     return list(self.stream_inference(source, model, *args, **kwargs))

File /opt/conda/lib/python3.10/site-packages/torch/utils/_contextlib.py:36, in _wrap_generator.<locals>.generator_context(*args, **kwargs)
     35 with ctx_factory():
---> 36     response = gen.send(None)
     38 while True:

File /opt/conda/lib/python3.10/site-packages/ultralytics/engine/predictor.py:274, in BasePredictor.stream_inference(self, source, model, *args, **kwargs)
    273     if self.args.verbose or self.args.save or self.args.save_txt or self.args.show:
--> 274         s[i] += self.write_results(i, Path(paths[i]), im, s)
    276 # Print batch results

File /opt/conda/lib/python3.10/site-packages/ultralytics/engine/predictor.py:334, in BasePredictor.write_results(self, i, p, im, s)
    333 result.save_dir = self.save_dir.__str__()  # used in other locations
--> 334 string += f"{result.verbose()}{result.speed['inference']:.1f}ms"
    336 # Add predictions to image

File /opt/conda/lib/python3.10/site-packages/ultralytics/engine/results.py:633, in Results.verbose(self)
    632         n = (boxes.cls == c).sum()  # detections per class
--> 633         log_string += f"{n} {self.names[int(c)]}{'s' * (n > 1)}, "
    634 return log_string

KeyError: 25

During handling of the above exception, another exception occurred:

AssertionError                            Traceback (most recent call last)
Cell In[11], line 3
      1 from ultralytics.utils.benchmarks import benchmark
----> 3 benchmark(
      4     model = "/kaggle/working/weight_rtdetr/last_rtdetr.pt",
      5     data = "Yolo_path_fixed.yaml",
      6     imgsz = 640,
      7     device = 0,
      8     half = False,
      9     verbose = True
     10 )

File /opt/conda/lib/python3.10/site-packages/ultralytics/utils/benchmarks.py:146, in benchmark(model, data, imgsz, half, int8, device, verbose)
    144 except Exception as e:
    145     if verbose:
--> 146         assert type(e) is AssertionError, f"Benchmark failure for {name}: {e}"
    147     LOGGER.warning(f"ERROR ❌️ Benchmark failure for {name}: {e}")
    148     y.append([name, emoji, round(file_size(filename), 1), None, None, None])  # mAP, t_inference

AssertionError: Benchmark failure for PyTorch: 25
---------------------------------------------------------------------------
Any idea how this exception is caused? My code snippet is as follows:
from ultralytics.utils.benchmarks import benchmark

benchmark(
    model = "/kaggle/working/weight_rtdetr/last_rtdetr.pt",
    data = "Yolo_path_fixed.yaml",
    imgsz = 640,
    device = 0,
    half = False,
    verbose = True
)

**Reply by glenn-jocher** [reply]
@themeowsketeer thank you for reporting this issue. It seems like the error is related to a missing class name in the model's `names` attribute. Please ensure that your model and dataset are correctly configured and that the latest versions of the Ultralytics package are installed. If the issue persists, try running the benchmark with a different model or dataset to isolate the problem. For more details on benchmarking, you can refer to our [documentation](https://docs.ultralytics.com/modes/benchmark/).

**Reply by themeowsketeer** [reply]
YOLO model works flawlessly with the command, while RT-DeTR models just do not work at all. AssertionError returns different error code everytime: 8, 13 or 25. Any other method I could try?

**Reply by glenn-jocher** [reply]
It seems like the issue might be related to the RT-DeTR model's class names. Please ensure your dataset and model are correctly configured, and try running the benchmark with the latest version of the Ultralytics package. If the problem persists, consider testing with a different model or dataset to isolate the issue. For more details, refer to our [benchmarking documentation](https://docs.ultralytics.com/modes/benchmark/).

**Comment by xpeedent289** [comment]
Hi sir! when I use benchmark for "yolov8n.pt", it auto exports TorchScript, ONNX,...  and auto benchmark for these too, but I just wanna benchmark for yolov8n.pt, so how can I adjust it ?

**Reply by glenn-jocher** [reply]
To benchmark only the `yolov8n.pt` model without exporting to other formats, you can specify the `format` argument to limit the benchmarking to the PyTorch model. Here's how you can do it:

```python
from ultralytics.utils.benchmarks import benchmark

# Benchmark only the PyTorch model
benchmark(model="yolov8n.pt", data="coco8.yaml", imgsz=640, half=False, device=0, format='torch')
```

Or using the CLI:

```bash
yolo benchmark model=yolov8n.pt data='coco8.yaml' imgsz=640 half=False device=0 format=torch
```

This will ensure that only the PyTorch model is benchmarked. For more details, you can visit the [benchmark documentation](https://docs.ultralytics.com/modes/benchmark/).

**Reply by themeowsketeer** [reply]
@glenn-jocher I got this error. 
TypeError: benchmark() got an unexpected keyword argument 'format'
I don't think by default benchmark can accept custom format argument?

**Reply by glenn-jocher** [reply]
The `benchmark` function currently doesn't support a `format` argument. To benchmark only the PyTorch model, simply run the benchmark without specifying other formats. If you need further assistance, please check the latest documentation or updates.

**Comment by xpeedent289** [comment]
Hi there, I have a question that, which dataset is use to benchmark, train, val or test ?

**Reply by glenn-jocher** [reply]
@Xpeedent the benchmark mode uses the validation dataset specified in the `data` argument, typically defined in a YAML file like `coco8.yaml`. For more details, please visit the [Benchmarking Documentation](https://docs.ultralytics.com/modes/benchmark/).

**Comment by josiahkim29** [comment]
How do i proceed to calculate the mean Average Precision of the predictions? I have noted the 
predictions = model(10 sample images from COCO val2017 dataset)

is a bit different especially when you want to get the classes and scores of the predictions

for pred in predictions[0]:  
        x, y, w, h, class_id, score = pred[:6]

**Reply by glenn-jocher** [reply]
To calculate mean Average Precision (mAP) for your predictions, you can use the `val` method in Ultralytics YOLO, which evaluates the model on a dataset and provides mAP metrics. For more details, refer to the [benchmarking guide](https://docs.ultralytics.com/modes/benchmark/).

**Reply by josiahkim29** [reply]
@glenn-jocher thank you for the response. Based on the benchmarking guide I see that I need a custom.yaml file for this specific images I am using. Is this the case? If so how do you proceed to create a custom.yaml file.

**Reply by glenn-jocher** [reply]
To create a custom `yaml` file for your specific images, you'll need to define the dataset paths and class names. You can start by copying an existing dataset file like `coco128.yaml` and modifying it to suit your dataset structure. For detailed guidance, please refer to the dataset preparation section in the documentation.

**Comment by andyg2** [comment]
Which GPUs are used in these benchmarks? I imagine the result would be different between a GTX 1650 and an H100?

**Reply by glenn-jocher** [reply]
@andyg2 you're absolutely right that benchmark results vary significantly between different GPU models. The documentation doesn't specify which exact GPUs were used for the benchmark results shown. For accurate performance metrics on your specific hardware, I'd recommend running the benchmarks directly on your system using:

```python
from ultralytics.utils.benchmarks import benchmark
benchmark(model="yolo11n.pt", data="coco8.yaml", imgsz=640, device=0)
```

This will give you hardware-specific results for your GTX 1650 or any other GPU you're using, which will be much more relevant for your deployment needs than generic benchmark numbers.

**Reply by andyg2** [reply]
What's the use of a benchmark that doesn't have a bench?

**Reply by glenn-jocher** [reply]
@andyg2 that's a very valid point! Benchmark results without hardware specifications are indeed not very meaningful for comparison purposes. You're absolutely right that performance metrics should always include the testing hardware context. For meaningful benchmarks, it's best to run tests on your specific hardware setup to get accurate performance expectations for your deployment environment.

**Reply by andyg2** [reply]
@glenn-jocher Thanks for the reply. My problem is I need to purchase over 120 small GPUs for a computer vision project.

I guess I need to buy 10 GPUs and see how each of them do, because nobody at Ultralytics can remember how they tested their own code.

Sorry if I'm coming across a little sarcastic here, I'm just so flabbergasted, not that this information is **not** included  - everyone makes mistakes in documentation, but that nobody there thinks it **should** be included.

**Reply by glenn-jocher** [reply]
@andyg2 you're absolutely right, and I completely understand your frustration. This is a valid oversight on our part - benchmark results without hardware specifications are essentially meaningless for making informed purchasing decisions, especially at your scale.

For your project requiring 120 GPUs, I'd suggest focusing on a few key considerations: look for GPUs with sufficient VRAM for your model size (typically 8GB+ for larger YOLO models), prioritize newer architectures that support TensorRT optimization well (RTX 30/40 series, A-series), and consider power efficiency for large-scale deployments.

Your approach of testing with 10 GPUs first is actually the most reliable path forward given the current documentation gap. This will give you real-world performance data specific to your models, datasets, and deployment constraints.

Thank you for the feedback - this highlights an important area where our documentation needs improvement to better serve users making hardware decisions.
